### DISCUSSION TITLE
curl behavior on error while proxying

### DISCUSSION THREAD

**Discussion by aathan** [curl/curl#9508] [Q&A] [answered]
I have proxy that returns a 503 with application/json content:

```
$ curl -o - -I -x 'http://x:y@localhost:1234' https://ipinfo.io/ip
HTTP/1.1 503 Service Unavailable
Content-Type: application/problem+json
Date: Thu, 15 Sep 2022 00:22:02 GMT
Content-Length: 214
```

However, curl does not show the content returned with the error. Is that a bug or some aspect of the HTTP spec that states such content should be ignored?

```
$ tcpdump -Ai lo port 1234
...
HTTP/1.1 503 Service Unavailable
Content-Type: application/problem+json
Date: Thu, 15 Sep 2022 00:22:02 GMT
Content-Length: 214

{"type":"https://xxx.com/problems/xxx-no-clients","title":"xxx","status":503,"detail":"xxx"}
```

... the content is definitely returned to curl...

**Comment by dfandrich** [comment, marked as answer]
The -I option that you're using, a.k.a. --head, explicitly requests that the 
body NOT be returned.

**Reply by aathan** [reply]
Removing the -I doesn't change the behavior, nor does including the -o (like I did). I have not found a way to get curl to emit the body of the error response when using -x. (I have not tried to see if I can get it to emit the body of an error response when *not* using -x.)

PS: I took that combination of params from a post I found while googling around for a solution, where it was suggested the -o would add the body back into the output.

**Reply by dfandrich** [reply]
Without the -I option curl should return the body, even for an HTTP error code. You're not showing the full tcpdump input so I can't tell how to interpret the output. Are you sure the proxy isn't eating the body? What is the output of curl --version?

**Reply by aathan** [reply]
I am the author of the proxy, and I ran across this problem while trying to debug it. Curl emits the 503 my code returns, and the tcpdump confirms that the 503 response returned by my proxy includes the body -- content length 214 etc (though I truncated the full text in what I posted here in order not to expose proprietary information).

I'm "quite sure" that curl is not emitting the body of the error response that is received by it (nothing other than that response has a 503 in it).

```
$ curl --version
curl 7.81.0 (x86_64-pc-linux-gnu) libcurl/7.81.0 OpenSSL/3.0.2 zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3 OpenLDAP/2.5.12
Release-Date: 2022-01-05
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets zstd
```

**Reply by aathan** [reply]
... there is really nothing prior to that in the tcpdump output that seems relevant, since all the rest of the conversation is the data outbound to the proxy. What you're seeing there is the totality of what was returned to curl.

This felt like a bug but I didn't want to open one without asking here first.

**Reply by dfandrich** [reply]
Running either of these commands against a public web server:

```
curl http://the-internet.herokuapp.com/status_codes/503
curl -x http://the-internet.herokuapp.com:80/status_codes/503 http://the-internet.herokuapp.com/status_codes/503
```
return the body of the 503 error code, as expected. However, this command, which is closer to yours:

```
curl -x http://the-internet.herokuapp.com:80/status_codes/503 https://the-internet.herokuapp.com/status_codes/503
```
does not, instead returning curl error code 56, "Received HTTP code 500 from proxy after CONNECT". The difference between this command and the previous is that the https: URL requested forces a CONNECT, and in that case the body is not returned. This looks to me to be deliberate, since a secure connection to the destination server could not be established so anything returned will not be from the desired server.

**Reply by bagder** [reply]
Correct, since `CONNECT`fails, the requested transfer could not be performed and the only body would be from the proxy and that is not what the transfer is about.
